GPTutils.py

```diff
@@ -0,0 +1,262 @@
+            value (tuple[str, str]): A tuple containing the role and content of the new message.
+        """
+        role, content = value
+        self.edit_message(index, role, content)
+
+    def __str__(self) -> str:
+        """
+        Returns a readable string representation of the conversation.
+
+        Returns:
+            str: A string representation of the conversation.
+        """
+        return "\n".join(
+            f"{msg['role']}: {msg['content']}" for msg in self.conversation
+        )
+
+
+def to_docs(text: str, file_extension: str) -> List[str]:
+    """
+    Splits the input text into a list of documents based on the file extension.
+
+    Args:
+    text (str): The input text to be split.
+    file_extension (str): The file extension of the input text.
+
+    Returns:
+    List[str]: A list of documents resulting from the split.
+    """
+
+    if file_extension in mapper:
+        language = mapper[file_extension]
+        python_splitter = RecursiveCharacterTextSplitter.from_language(
+            language=language, chunk_size=500, chunk_overlap=0
+        )
+        docs = python_splitter.create_documents([text])
+        docs = [doc.page_content for doc in docs]
+    else:
+        text_splitter = CharacterTextSplitter(
+            separator="\n\n",
+            chunk_size=500,
+            chunk_overlap=0,
+            length_function=len,
+            is_separator_regex=False,
+        )
+        docs = text_splitter.split_text(text)
+    return docs
+
+
+def cosim(a, b):
+    """
+    Computes the cosine similarity between two vectors a and b.
+
+    Args:
+    a (numpy.ndarray): The first vector.
+    b (numpy.ndarray): The second vector.
+
+    Returns:
+    float: The cosine similarity between a and b.
+    """
+    return np.dot(a, b) / (norm(a) * norm(b))
+
+
+def cosim_matrix(a, b):
+    """
+    Computes the cosine similarity matrix between two matrices a and b.
+
+    Args:
+    a (numpy.ndarray): An array of shape (n).
+    b (numpy.ndarray): A matrix of shape (k, n).
+
+    Returns:
+    numpy.ndarray: An array of shape (k) containing the cosine similarity between each row of a and b.
+
+    Examples:
+    >>> a = np.array([1, 2, 3])
+    >>> b = np.array([[1, 2, 3], [4, 5, 6]])
+    >>> cosim_matrix(a, b)
+    array([1.        , 0.97463185])
+    """
+    return np.dot(a, b.T) / (norm(a) * norm(b, axis=1))
+
+
+class FilesIndex:
+    def __init__(self, files: dict[str, str]) -> None:
+        """
+        Initializes a new instance of the FilesIndex class.
+
+        Args:
+            files (dict[str, str]): A dictionary containing the files to be indexed.
+            {Filename: FileContent}
+        """
+        self.files = files
+        self.index: Dict[str, List[int]] = {}
+        self.idx_to_filename = {}
+        self.embeddings_model = OpenAIEmbeddings()
+        self.docs: List[str] = []
+        self.embeddings: np.ndarray = None
+        self.index_files()
+
```

Comment:
Please make sure to add appropriate error handling to newly introduced functions in case of unexpected behaviors, for instance in asynchronous calls.
```suggestion
    def cosim(a, b):
        try:
            return np.dot(a, b) / (norm(a) * norm(b))
        except Exception as e:
            logging.error(f"Failed to calculate cosine similarity: {str(e)}")
            return 0.0
```